src/cpp/src/llm_pipeline.cpp

```diff
@@ -0,0 +1,353 @@
+    const std::string& path, 
+    const std::string& device, 
+    const ov::AnyMap& config, 
+    const std::string& ov_tokenizers_path
+): 
+    m_model_runner{ov::Core{}.compile_model(path + "/openvino_model.xml", device, config).create_infer_request()}, 
+    m_tokenizer{Tokenizer(path, device, ov_tokenizers_path)},
+    m_generation_config{from_config_json_if_exists(path)},
+    m_chat_template{from_tokenizer_json_if_exists(path)}
+ {}
+
+ov::genai::GenerationConfig ov::genai::LLMPipeline::LLMPipelineImpl::generation_config() const {
+    return m_generation_config;
+}
+
+ov::genai::GenerationConfig ov::genai::LLMPipeline::get_generation_config() const {
+    return m_pimpl->generation_config();
+}
+
+std::string ov::genai::LLMPipeline::LLMPipelineImpl::generate(
+    std::string text, 
+    OptionalGenerationConfig generation_config,
+    OptionalStreamerVariant streamer
+) {
+    GenerationConfig config = (generation_config.has_value()) ? *generation_config : m_generation_config;
+
+    if (is_chat_conversation) {
+        text = apply_chat_template(text);
+    }
+    auto kv_cache_len = m_model_runner.query_state()[0].get_state().get_shape()[2];
+    
+    // previous prompt generation in chat dialog stops with the end of sentence token, 
+    // need to append this token to the current prompt
+    if (is_chat_conversation && kv_cache_len > 0) {
+        text = config.eos_token + text;
+    }
+
+    auto [input_ids, attention_mask] = m_tokenizer.encode(text);
+
+    // todo: W/A If sentence begins with a specfial tokens (<bos>, <s>, etc.) openvino_tokenizer inserts 2 special extra tokens <bos> and "▁",
+    // but HF does not do that. Moreover openvino_tokenizer always inserts <bos> but in chat scenario HF does not do that because skip_special_tokens=True.
+    // Need to remove both of that tokens manually to get exact token by token alignment with HF
+    auto size = input_ids.get_shape();
+    int64_t* inputs_data = input_ids.data<int64_t>();
+    std::vector<int64_t> tmp_ids(inputs_data, inputs_data + input_ids.get_size()); // todo: works only for batch 1
+    // tmp_ids.erase(tmp_ids.begin());
+
+    auto attention_mask_data = attention_mask.data<int64_t>();
+    std::vector<float> tmp_attn_mask(attention_mask_data, attention_mask_data + attention_mask.get_size());
+    // tmp_attn_mask.erase(tmp_attn_mask.begin());
+
+    std::vector<std::string> prefixes_to_exclude = {config.eos_token, config.bos_token};
+    auto prefix_match = [&text](std::string prefix) { return text.substr(0, prefix.length()) == prefix; };
+    if (std::any_of(prefixes_to_exclude.begin(), prefixes_to_exclude.end(), prefix_match)) {
+        tmp_ids.erase(tmp_ids.begin());
+        tmp_attn_mask.erase(tmp_attn_mask.begin());
+    }
+
+    input_ids = ov::Tensor(input_ids.get_element_type(), {1, tmp_ids.size()});
+    for (size_t i = 0; i < tmp_ids.size(); i++)
+        input_ids.data<int64_t>()[i] = tmp_ids.data()[i];
```

Comment:
we don't need to perform copy, we can create Tensor on existing `tmp_ids` memory

Reply:
unfortunately, when i do i get segmentation fault error
```
input_ids = ov::Tensor{input_ids.get_element_type(), {1, tmp_ids.size()}, tmp_ids.data()};
attention_mask = ov::Tensor{attention_mask.get_element_type(), {1, tmp_attn_mask.size()}, tmp_attn_mask.data()};
```